udev: introduce udev net_id "naming schemes"

poettering opened the pull request:
With this we can stabilize how naming works for network interfaces. A
user can request through a kernel cmdline option or an env var which
scheme to follow. The idea is that installers use this to set into stone
(a very soft stone though) the scheme used during installation so that
interface naming doesn't change afterwards anymore.

Why use env vars and kernel cmdline options, and not a config file of
its own?

Well, first of all there's no obvious existing one to use. But more
importantly: I have the feeling that this logic is kind of an incomplete
hack, and I simply don't want to do advertise this as a perfectly
working solution. So far we used env vars for the non-so-official
options and proper config files for the official stuff. Given how
incomplete this logic is (i.e. the big variable for naming remains the
kernel, which might expose sysfs attributes in newer versions that we
check for and didn't exist in older versions — and other problems like
this), I am simply not confident in giving this first-class exposure in
a primary configuration file.

Fixes: #10448

poettering commented:
This probably shouldn't be merged the way this is (for example, it lacks doc additions). It's primarily just a proposal how to do things to fix #10448 to get the discussion started.

@keszybz can you comment?

poettering commented:
/cc @msekletar @lnykryn

keszybz commented on src/udev/udev-builtin-net_id.c:
```diff
@@ -150,6 +185,54 @@ struct virtfn_info {
+                                }
+
+                        if (!scheme)
+                                log_warning("Naming scheme '%s' requested but not defined. Ignoring.", k);
```
"Unknown naming scheme "%s" requested, ignoring" ?

keszybz commented on src/udev/udev-builtin-net_id.c:
```diff
@@ -150,6 +185,54 @@ struct virtfn_info {
+                }
+
+                if (scheme)
+                        log_info("Using selected interface naming scheme '%s'.", scheme->name);
```
"selected" → "" ?

poettering commented:
>  I'm also pretty sure we should provide a config-file inteface for this. `NamePolicyScheme=` in .link files seems natural.

This would require major reworking and API breakage: currently the potential stable names are determined early (`75-net-description.rules`), and the `.link` files are applied much later (`80-net-setup-link.rules`). And inbetween other rules can modify the names, replace them, extend them, remove them. Hence, `.link` files are not really a suitable place to configure anything like this, as it means we'd have to parse them twice, match them twice, once early and once late, and the keys that match might have changed in between.

So I am very sure that `.link` files are not the right place. They should be used for consuming the names determined, not for generating them.

keszybz commented:
OK, so if .link files are not good, then what about making a new file, e.g. `/etc/systemd/network/naming.conf` ?

keszybz commented on src/udev/udev-builtin-net_id.c:
```diff
@@ -115,6 +116,40 @@
+ * do, and thus affect our naming scheme too. Thus, enforcing a naming scheme will make interfacing more stable across
+ * OS versions, but not fully stabilize them. */
+typedef enum NamingSchemeFlags {
+        NAMING_INFINIBAND      = 1 << 0, /* Use "ib" prefix for infiniband, see 938d30aa98df887797c9e05074a562ddacdcdf5e */
```
The four commits mentioned on the right are
v239-861-g938d30aa98
v238-251-g609948c704
v238-252-g6bc04997b6
v239-1084-gd81186ef4f

I think they should be sorted by age.

keszybz commented on src/udev/udev-builtin-net_id.c:
```diff
@@ -115,6 +116,40 @@
+
+static const NamingScheme naming_schemes[] = {
+        { "v238", 0                                                                        },
+        { "v239", NAMING_INFINIBAND|NAMING_SR_IOV_V|NAMING_NPAR_ARI|NAMING_ZERO_ACPI_INDEX },
```
Actually this should be something like:
```C
v238, 0,
v239, NAMING_SR_IOV_V|NAMING_NPAR_ARI
v240, NAMING_SR_IOV_V|NAMING_NPAR_ARI|NAMING_INFINIBAND|NAMING_ZERO_ACPI_INDEX
```

poettering replied:
I think this is misleading. iiuc all four of those commits got merged post-239, even though the branches themselves were on top of 238, or did i misunderstand something there?

(we can later on add more entries for older versions btw. And downstreams can entries like "rhel7.1" or so too, if they like)

And yes, the growing linearly thing is an issue i see too, but i wanted to allow that new versions might be able to drop old mistakes again...

keszybz replied:
```console
$ git tag --contains 609948c7043a40008b8299529c978ed8e11de8f6
v239
```
NEWS also lists this as the first change for v239. The same for NPAR.

(The comment about linearity wasn't a complaint, just a note.)

poettering commented:
do we actually need a config file at all for this? i mean, for the net.ifnames= stuff we didn't have that either, and noone complained... In particular given your suggestion to just extend net.ifnames= maybe we should add exactly nothing new otherwise?

As mentioned, I am not a fan of using "offical" looking config files for something that is such a hacky, incomplete solution. Moreover, having a config file for a single setting, in particular one that looks so so much like .link files in their effect, and where everybody will just ask "why are they not .link files anyway?" the way you did, sounds bad to me too....

hence, i can drop the env var, and move everything to just the kernel cmdline option (even though overloading net.ifnames= like this is slightly ugly to me too, since it too gets currently interpreted only in the late step, not the early step, but we'd have to interpret now in the early step too. but for kernel cmdline vars this is less ugly than for config files with [Match] sections i guess)

keszybz commented:
I think the env var stuff is very nice for debugging, and I think it makes a lot of sense to keep it, at least for that purpose. OK, so let's skip the question of the config file for now. We certainly need a kernel-cmdline interface for support overriding in the initramfs, but the other interfaces are optional. I liked the idea of being able to specify the naming scheme versions independently for different files, but this seems infeasible due to the reasons you give, so indeed, a config file doesn't seems necessary.

Likewise, because this is interpreted earlier than `net.ifnames=`, it shouldn't be the same variable.

OK, so whatabout calling it just `net.naming-scheme=` ? This matches `net.ifnames=` nicely and it short enough to type from memory.

keszybz commented on src/udev/udev-builtin-net_id.c:
```diff
@@ -150,6 +185,54 @@ struct virtfn_info {
+                _cleanup_free_ char *buffer = NULL;
+                const char *e, *k;
+
+                /* Acquire setting form the kernel command line */
```
form → from

keszybz commented:
I'm setting the needs-rework flag because of the 238 vs. 239 vs. v240 mixup.

poettering commented:
I have force-pushed a new version now. It addresses all points made, and adds documentation. I also added a udev device property indicating the naming scheme actually used.
